server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala

```diff
@@ -114,6 +119,119 @@ class DeltaSharedTableKernel(
     (table, engine)
   }
 
+  // Get the table snapshot of the specified version or timestamp
+  // TODO: Validate the protocol and metadata
+  private def getSharedTableSnapshot(
+      table: Table,
+      engine: Engine,
+      version: Option[Long],
+      timestamp: Option[String],
+      validStartVersion: Long,
+      clientReaderFeatures: Set[String],
+      flagReaderFeatures: Set[String] = Set.empty,
+      isProviderRpc: Boolean = false): SharedTableSnapshot = {
+    val snapshot =
+      if (version.isDefined) {
+        try {
+          table.getSnapshotAsOfVersion(engine, version.get).asInstanceOf[SnapshotImpl]
+        } catch {
+          case e: io.delta.kernel.exceptions.TableNotFoundException =>
+            throw new FileNotFoundException(e.getMessage)
+        }
+
+      } else if (timestamp.isDefined) {
+        try {
+          table
+            .getSnapshotAsOfTimestamp(engine, millisSinceEpoch(timestamp.get))
+            .asInstanceOf[SnapshotImpl]
+        } catch {
+          // Convert to DeltaSharingIllegalArgumentException to return 4xx instead of 5xx error code
+          // Only convert known exceptions around timestamp too late or too early
+          case e: io.delta.kernel.exceptions.TableNotFoundException =>
+            throw new FileNotFoundException(e.getMessage)
+        }
+      } else {
+        val classLoader = this.getClass.getClassLoader
+        Thread.currentThread().setContextClassLoader(classLoader)
+        try {
+          table.getLatestSnapshot(engine).asInstanceOf[SnapshotImpl]
+        }
+        catch {
+          case e: io.delta.kernel.exceptions.TableNotFoundException =>
+            throw new FileNotFoundException(e.getMessage)
+          case e: Throwable =>
+            throw new FileNotFoundException(e.getMessage)
```

Comment:
@linzhou-db I needed to catch all throwabales here to fix the issue in this test case: table_changes_exception: cdf_table_missing_log from DeltaSharingSuite.scala. I get a runtime error for this test case on the table.getLatestSnapshot(engine).asInstanceOf[SnapshotImpl] line and it needs to be handled as a 400 error and not a 500 error so I need to throw a FileNotFound, but I could not figure out what I need to catch, and so I caught all throwables.

Reply:
doing string matching on error message now since fix is not in delta-kernel yet